{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "44a46434-7fea-4ffd-8f80-09ed518767b3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-23T16:04:23.448214Z",
     "iopub.status.busy": "2024-06-23T16:04:23.447811Z",
     "iopub.status.idle": "2024-06-23T16:04:23.457693Z",
     "shell.execute_reply": "2024-06-23T16:04:23.457097Z",
     "shell.execute_reply.started": "2024-06-23T16:04:23.448196Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import ks_2samp, entropy\n",
    "from scipy.signal import convolve\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "from scipy.spatial.distance import jensenshannon\n",
    "\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s',\n",
    "    filename='methylation_smoothing.log', \n",
    "    filemode='w'                        \n",
    ")\n",
    "\n",
    "def calculate_divergences_and_ks(df, smoothing_method=\"none\", window_size=5, smoothing_factor=1e-5, frac=0.1):\n",
    "    \"\"\"Calculates normalized divergences (JSD, KL, GJS, SGJS) and KS test for methylation data with optional smoothing.\n",
    "\n",
    "    Args:\n",
    "        df (pandas.DataFrame): DataFrame with 'methylated' and 'unmethylated' columns.\n",
    "        smoothing_method (str, optional): Smoothing method to apply. Options: \"none\", \"moving_average\", \"lowess\", \"gaussian\", \"laplace\". Defaults to \"none\".\n",
    "        window_size (int, optional): Window size for moving average smoothing. Defaults to 5.\n",
    "        smoothing_factor (float, optional): Smoothing factor for Laplace smoothing (add to numerator and denominator). Defaults to 1e-5.\n",
    "        frac (float, optional): Fraction of data to use for LOWESS smoothing. Defaults to 0.1.\n",
    "\n",
    "    Returns:\n",
    "        tuple: Normalized divergences (JSD, KL, GJS, SGJS) and KS test statistic and p-value.\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "     # Filter and drop rows as before\n",
    "    df = df[(df['methylated'] != 0) | (df['unmethylated'] != 0)].copy()  \n",
    "    if len(df) == 0:\n",
    "        return None, None, None, None, None, None\n",
    "\n",
    "    total = df[\"methylated\"] + df[\"unmethylated\"]\n",
    "    df[\"methylation_ratio\"] = df[\"methylated\"] / total\n",
    "    \n",
    "    # Replace inplace with this line\n",
    "    df[\"methylation_ratio\"] = df[\"methylation_ratio\"].fillna(0)  \n",
    "\n",
    "    \n",
    "    # Apply smoothing based on the selected method\n",
    "    if smoothing_method == \"moving_average\":\n",
    "        window = np.ones(window_size) / window_size\n",
    "        smoothed_ratios = convolve(df[\"methylation_ratio\"], window, mode=\"same\")\n",
    "        df[\"smoothed_methylation_ratio\"] = smoothed_ratios.clip(0, 1) \n",
    "        \n",
    "    elif smoothing_method == \"lowess\":\n",
    "        smoothed_ratios = lowess(df[\"methylation_ratio\"], np.arange(len(df)), frac=frac)[:, 1]\n",
    "        df[\"smoothed_methylation_ratio\"] = smoothed_ratios.clip(0, 1)  \n",
    "        \n",
    "    elif smoothing_method == \"gaussian\":\n",
    "        smoothed_ratios = gaussian_filter1d(df[\"methylation_ratio\"], sigma=window_size)\n",
    "        df[\"smoothed_methylation_ratio\"] = smoothed_ratios.clip(0, 1)  \n",
    "        \n",
    "    elif smoothing_method == \"laplace\":\n",
    "        df[\"smoothed_methylation_ratio\"] = (df[\"methylated\"] + smoothing_factor) / (total + 2 * smoothing_factor) \n",
    "        \n",
    "    else:  # No smoothing\n",
    "        df[\"smoothed_methylation_ratio\"] = df[\"methylation_ratio\"]  \n",
    "\n",
    "    # Divergence calculations using smoothed ratios\n",
    "    p = df[\"smoothed_methylation_ratio\"]\n",
    "    q = 1 - p\n",
    "\n",
    " # Divergence calculations (with normalizations)\n",
    "    normalized_js_divergence = float(jensenshannon(p, q, base=2))  # JSD\n",
    "\n",
    "    normalized_kl_divergence = float(entropy(p, q, base=2) - entropy(p, base=2))  # KLD\n",
    "\n",
    "    # Geometric Jensen-Shannon Divergence\n",
    "    normalized_gjs_divergence = float(np.sqrt(\n",
    "        0.5 * (np.sqrt(entropy(p, q, base=2)) + np.sqrt(entropy(q, p, base=2)))\n",
    "    ) / np.sqrt(np.log(2)))\n",
    "    \n",
    "    # Symmetric Geometric Jensen-Shannon Divergence\n",
    "    m = (p + q) / 2\n",
    "    normalized_sgjs_divergence = float( 0.5 * (\n",
    "        np.sqrt(jensenshannon(p, m, base=2)) + np.sqrt(jensenshannon(q, m, base=2))\n",
    "    ) / np.sqrt(np.log(2)))\n",
    "\n",
    "    # Kolmogorov-Smirnov Test\n",
    "    ks_statistic, ks_pvalue = ks_2samp(df[\"methylated\"], df[\"unmethylated\"])  # KST\n",
    "    \n",
    "\n",
    "    return normalized_js_divergence, normalized_kl_divergence, normalized_gjs_divergence, normalized_sgjs_divergence, ks_statistic, ks_pvalue\n",
    "\n",
    "\n",
    "def read_bismark_file(filename):\n",
    "    \"\"\"Reads Bismark methylation data from a file.\"\"\"\n",
    "    column_names = [\"chr\", \"start\", \"end\", \"coverage\", \"methylated\", \"unmethylated\"]\n",
    "    df = pd.read_csv(filename, sep='\\t', header=None, names=column_names, compression='gzip')\n",
    "\n",
    "    # Convert columns to correct data types\n",
    "    df['start'] = pd.to_numeric(df['start'])\n",
    "    df['end'] = pd.to_numeric(df['end'])\n",
    "    df['coverage'] = pd.to_numeric(df['coverage'])\n",
    "    df['methylated'] = pd.to_numeric(df['methylated'])\n",
    "    df['unmethylated'] = pd.to_numeric(df['unmethylated'])\n",
    "\n",
    "    return df\n",
    "\n",
    "def clean_data(results_df):\n",
    "    \"\"\"Cleans the results DataFrame by removing file extensions from sample names.\"\"\"\n",
    "    results_df['Sample'] = results_df['Sample'].astype(str).str.replace('.bedgraph.gz', '', regex=False)\n",
    "    return results_df\n",
    "\n",
    "\n",
    "\n",
    "# Data processing pipeline\n",
    "data_directory = \"/shares/grossniklaus.botinst.uzh/dkt/projects/meth1000/analysis/09_split_cov_chr/output\"\n",
    "results = []\n",
    "\n",
    "# Get all sample subdirectories\n",
    "sample_directories = [d for d in os.listdir(data_directory) if os.path.isdir(os.path.join(data_directory, d))]\n",
    "\n",
    "\n",
    "# Iterate through files\n",
    "for sample_dir in sample_directories:\n",
    "    sample_path = os.path.join(data_directory, sample_dir)\n",
    "\n",
    "    logging.info(f\"Processing sample directory: {sample_path}\")\n",
    "    print(sample_path)\n",
    "\n",
    "    # Iterate through files in each sample directory\n",
    "    for file_path in glob.glob(os.path.join(sample_path, \"*_chr_*.cov.gz\")):\n",
    "        try:\n",
    "            \n",
    "            filename = os.path.basename(file_path)\n",
    "            parts = filename.split(\"_\")\n",
    "\n",
    "            context_type = parts[0]\n",
    "            sample_name = parts[1] + \"_\" + parts[2].split(\".\")[0]\n",
    "            chromosome = parts[-1].split(\".\")[0]\n",
    "\n",
    "            df = read_bismark_file(file_path)\n",
    "\n",
    "            smoothing_methods = [\"none\", \"moving_average\", \"lowess\", \"gaussian\", \"laplace\"]\n",
    "            smoothing_results = {}\n",
    "\n",
    "            for smoothing_method in smoothing_methods:\n",
    "\n",
    "                divergences = calculate_divergences_and_ks(df, smoothing_method, window_size=3)\n",
    "\n",
    "                smoothing_results[smoothing_method] = {\n",
    "                    \"Divergences\": divergences,\n",
    "                    \"Methylation Ratio\": df[\"methylation_ratio\"],\n",
    "                    \"Smoothed Methylation Ratio\": df[\"smoothed_methylation_ratio\"] if smoothing_method != \"none\" else None  \n",
    "                }\n",
    "\n",
    "            # Plotting\n",
    "            plt.figure(figsize=(12, 6))\n",
    "            plt.plot(df[\"methylation_ratio\"], label=\"Original\", alpha=0.5)\n",
    "\n",
    "            for smoothing_method, data in smoothing_results.items():\n",
    "                if data[\"Smoothed Methylation Ratio\"] is not None:\n",
    "                    plt.plot(data[\"Smoothed Methylation Ratio\"], label=smoothing_method)\n",
    "\n",
    "            plt.title(f\"Methylation Ratios with Different Smoothing Methods - Sample: {sample_name} - Chromosome: {chromosome}\")\n",
    "            plt.xlabel(\"Position\")\n",
    "            plt.ylabel(\"Methylation Ratio\")\n",
    "            plt.legend()\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(f\"/path/to/your/plot/directory/{sample_name}_{chromosome}_{context_type}_smoothing_comparison.png\")\n",
    "            plt.close()\n",
    "    \n",
    "            results.append(\n",
    "                {\n",
    "                    \"Sample\": sample_name,\n",
    "                    \"Chromosome\": chromosome,\n",
    "                    \"Context\": context_type,\n",
    "                    **{\n",
    "                        f\"{method} Divergence\": smoothing_results[method][\"Divergences\"][i] \n",
    "                        for method in smoothing_methods\n",
    "                        for i in range(len(smoothing_results[method][\"Divergences\"]))  # Iterate over all divergence results\n",
    "                    }\n",
    "                }\n",
    "            )\n",
    "        \n",
    "\n",
    "\n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error processing file {filename}: {e}\")\n",
    "\n",
    "# Create and save results DataFrame\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df.to_csv(\"/shares/grossniklaus.botinst.uzh/eharputluoglu/calculation_output/methylation_analysis_results_Smooting_all.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f04e71-6323-4992-a140-a040796176ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5cb5af7-3386-48d4-83f0-c97eb1ff2708",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dbe9848-44da-4f91-8264-724224fcfcef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2dc690b-9152-4a93-97ca-dc43d0197fab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a8ce265-8e9e-435f-9a26-23d855e609de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c2b24dd-15e4-47fa-a0c0-9fc583a0e53b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0193b467-e216-436a-9362-7903adbd2ce3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fc47c072-36d9-407c-b559-89a09e7e0e60",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### OLDDDD\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d6825c5-d2f3-42d4-802d-f7f21c95a3c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def read_bismark_file(filename):\n",
    "    \"\"\"Reads Bismark methylation data from a file.\"\"\"\n",
    "    column_names = [\"chr\", \"start\", \"end\", \"coverage\", \"methylated\", \"unmethylated\"]\n",
    "    df = pd.read_csv(filename, sep='\\t', header=None, names=column_names, compression='gzip')\n",
    "\n",
    "    # Convert columns to correct data types\n",
    "    df['start'] = pd.to_numeric(df['start'])\n",
    "    df['end'] = pd.to_numeric(df['end'])\n",
    "    df['coverage'] = pd.to_numeric(df['coverage'])\n",
    "    df['methylated'] = pd.to_numeric(df['methylated'])\n",
    "    df['unmethylated'] = pd.to_numeric(df['unmethylated'])\n",
    "\n",
    "    return df\n",
    "\n",
    "def clean_data(results_df):\n",
    "    \"\"\"Cleans the results DataFrame by removing file extensions from sample names.\"\"\"\n",
    "    results_df['Sample'] = results_df['Sample'].astype(str).str.replace('.bedgraph.gz', '', regex=False)\n",
    "    return results_df\n",
    "\n",
    "\n",
    "# Data processing pipeline\n",
    "data_directory = \"/shares/grossniklaus.botinst.uzh/dkt/projects/meth1000/analysis/09_split_cov_chr/output\"\n",
    "results = []\n",
    "\n",
    "# Get all sample subdirectories\n",
    "sample_directories = [d for d in os.listdir(data_directory) if os.path.isdir(os.path.join(data_directory, d))]\n",
    "\n",
    "# Iterate through files\n",
    "for sample_dir in sample_directories:\n",
    "    sample_path = os.path.join(data_directory, sample_dir)\n",
    "\n",
    "    logging.info(f\"Processing sample directory: {sample_path}\")\n",
    "\n",
    "    # Iterate through files in each sample directory\n",
    "    for file_path in glob.glob(os.path.join(sample_path, \"*_chr_*.cov.gz\")):\n",
    "        try:\n",
    "            filename = os.path.basename(file_path)\n",
    "            parts = filename.split(\"_\")\n",
    "\n",
    "            context_type = parts[0]\n",
    "            sample_name = parts[1] + \"_\" + parts[2].split(\".\")[0] \n",
    "            chromosome = parts[-1].split(\".\")[0]\n",
    "\n",
    "            #logging.info(f\"Processing file: {filename}\")\n",
    "\n",
    "\n",
    "\n",
    "            df = read_bismark_file(file_path)\n",
    "            divergences = calculate_divergences_and_ks(df, smoothing_method=\"gaussian\", window_size=3)\n",
    "    \n",
    "            results.append(\n",
    "                {\n",
    "                    \"Sample\": sample_name,\n",
    "                    \"Chromosome\": chromosome,\n",
    "                    \"Context\": context_type,\n",
    "                    \"JS Divergence\": divergences[0],  \n",
    "                    \"KL Divergence\": divergences[1],\n",
    "                    \"GJS Divergence\": divergences[2],\n",
    "                    \"SGJS Divergence\": divergences[3],\n",
    "                    \"KS Statistic\": divergences[4],\n",
    "                    \"KS P-value\": divergences[5]\n",
    "                }\n",
    "            )\n",
    "    \n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error processing file {filename}: {e}\")\n",
    "\n",
    "# Create a DataFrame from results\n",
    "results_df = pd.DataFrame(results)\n",
    "\n",
    "# Clean data\n",
    "#results_df = clean_data(results_df)\n",
    "\n",
    "# Save the results to a CSV file\n",
    "results_df.to_csv(\"/shares/grossniklaus.botinst.uzh/eharputluoglu/calculation_output/methylation_analysis_results_Smooting_1e-5.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5180fff6-64d4-4623-860a-8d389e6a8246",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-11T13:38:45.207452Z",
     "iopub.status.busy": "2024-06-11T13:38:45.206933Z",
     "iopub.status.idle": "2024-06-11T13:38:45.218061Z",
     "shell.execute_reply": "2024-06-11T13:38:45.217709Z",
     "shell.execute_reply.started": "2024-06-11T13:38:45.207432Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "03c0e627-eee4-41c8-ba72-c5bcb24583c3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-11T14:04:45.188443Z",
     "iopub.status.busy": "2024-06-11T14:04:45.188078Z",
     "iopub.status.idle": "2024-06-11T14:04:45.439860Z",
     "shell.execute_reply": "2024-06-11T14:04:45.439446Z",
     "shell.execute_reply.started": "2024-06-11T14:04:45.188425Z"
    }
   },
   "outputs": [],
   "source": [
    "column_names = [\"chr\", \"start\", \"end\", \"coverage\", \"methylated\", \"unmethylated\"]\n",
    "df = pd.read_csv(\"/shares/grossniklaus.botinst.uzh/eharputluoglu/test_run_datasets/SRX1664464_se/CHG_SRX1664464_se_chr2.bismark.cov.gz\", sep='\\t', header=None, names=column_names, compression='gzip')\n",
    "\n",
    "# Convert columns to correct data types\n",
    "df['start'] = pd.to_numeric(df['start'])\n",
    "df['end'] = pd.to_numeric(df['end'])\n",
    "df['coverage'] = pd.to_numeric(df['coverage'])\n",
    "df['methylated'] = pd.to_numeric(df['methylated'])\n",
    "df['unmethylated'] = pd.to_numeric(df['unmethylated'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47a3e864-a4f6-4331-8d01-6c0a1338953a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-11T14:04:45.619377Z",
     "iopub.status.busy": "2024-06-11T14:04:45.618873Z",
     "iopub.status.idle": "2024-06-11T14:04:45.626136Z",
     "shell.execute_reply": "2024-06-11T14:04:45.625842Z",
     "shell.execute_reply.started": "2024-06-11T14:04:45.619363Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>coverage</th>\n",
       "      <th>methylated</th>\n",
       "      <th>unmethylated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1013</td>\n",
       "      <td>1014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1034</td>\n",
       "      <td>1035</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1084</td>\n",
       "      <td>1085</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>1087</td>\n",
       "      <td>1088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>1108</td>\n",
       "      <td>1109</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849857</th>\n",
       "      <td>2</td>\n",
       "      <td>19697521</td>\n",
       "      <td>19697522</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849858</th>\n",
       "      <td>2</td>\n",
       "      <td>19697523</td>\n",
       "      <td>19697524</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849859</th>\n",
       "      <td>2</td>\n",
       "      <td>19697536</td>\n",
       "      <td>19697537</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849860</th>\n",
       "      <td>2</td>\n",
       "      <td>19697538</td>\n",
       "      <td>19697539</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849861</th>\n",
       "      <td>2</td>\n",
       "      <td>19697784</td>\n",
       "      <td>19697785</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>849862 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end  coverage  methylated  unmethylated\n",
       "0         2      1013      1014       0.0           0             1\n",
       "1         2      1034      1035       0.0           0             2\n",
       "2         2      1084      1085       0.0           0             2\n",
       "3         2      1087      1088       0.0           0             1\n",
       "4         2      1108      1109       0.0           0             1\n",
       "...     ...       ...       ...       ...         ...           ...\n",
       "849857    2  19697521  19697522       0.0           0             4\n",
       "849858    2  19697523  19697524       0.0           0             8\n",
       "849859    2  19697536  19697537       0.0           0             3\n",
       "849860    2  19697538  19697539       0.0           0             5\n",
       "849861    2  19697784  19697785      20.0           1             4\n",
       "\n",
       "[849862 rows x 6 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25022edf-652a-4776-997b-1259aeaba19d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "environment_shannon",
   "language": "python",
   "name": "environment_shannon"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
