{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c17f8158-49e4-44cb-bfd7-056377eddf88",
   "metadata": {
    "editable": true,
    "execution": {
     "iopub.execute_input": "2024-06-25T14:17:39.056729Z",
     "iopub.status.busy": "2024-06-25T14:17:39.055818Z",
     "iopub.status.idle": "2024-06-25T14:17:39.067240Z",
     "shell.execute_reply": "2024-06-25T14:17:39.066451Z",
     "shell.execute_reply.started": "2024-06-25T14:17:39.056704Z"
    },
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import entropy, ks_2samp\n",
    "from scipy.special import rel_entr\n",
    "import json\n",
    "import zipfile\n",
    "import os\n",
    "import glob\n",
    "import os\n",
    "import logging\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.spatial.distance import jensenshannon\n",
    "from scipy.stats import entropy, ks_2samp, f_oneway\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "from scipy.special import kl_div\n",
    "\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s',\n",
    "    filename='methylation_analysis_3.log', \n",
    "    filemode='w'                        \n",
    ")\n",
    "\n",
    "def divergence_calculations(data, each_row=True, smoothing=1e-10):\n",
    "    # Function to calculate KL divergence\n",
    "    def kl_divergence(p, q):\n",
    "        return np.sum(rel_entr(p, q))\n",
    "\n",
    "    # Function to calculate Geometric Jensen-Shannon Divergence\n",
    "    def geometric_jsd(p, q):\n",
    "        jsd = jensenshannon(p, q)\n",
    "        return np.sqrt(jsd)\n",
    "    # Function to calculate Kolmogorov-Smirnov Test\n",
    "    def ks_test(p, q):\n",
    "        return ks_2samp(p, q).statistic\n",
    "\n",
    "    results = []\n",
    "\n",
    "    if each_row:\n",
    "        for index, row in data.iterrows():\n",
    "            p = np.array([row['methylated'], row['unmethylated']], dtype=np.float64)\n",
    "            q = np.array([row['unmethylated'], row['methylated']], dtype=np.float64)\n",
    "            \n",
    "            # Apply smoothing\n",
    "            p += smoothing\n",
    "            q += smoothing\n",
    "            \n",
    "            # Normalize p and q\n",
    "            p /= p.sum()\n",
    "            q /= q.sum()\n",
    "            \n",
    "            kl = kl_divergence(p, q)\n",
    "            js = jensenshannon(p, q)\n",
    "            gjs = geometric_jsd(p, q)\n",
    "            ks = ks_test(p, q)\n",
    "\n",
    "            row_result = {\n",
    "                'index': index,\n",
    "                'entropy': entropy(p),\n",
    "                'relative_entropy': kl,  # KL divergence is relative entropy\n",
    "                'dkl_divergence': kl,\n",
    "                'jsd': js,\n",
    "                'geometric_jsd': gjs,\n",
    "                'kolmogorov_smirnov': ks\n",
    "            }\n",
    "\n",
    "            results.append(row_result)\n",
    "\n",
    "        # Save the results to a JSON file\n",
    "        with open('results.json', 'w') as f:\n",
    "            json.dump(results, f)\n",
    "        \n",
    "        # Zip the JSON file\n",
    "        with zipfile.ZipFile('results.zip', 'w') as zipf:\n",
    "            zipf.write('results.json')\n",
    "        \n",
    "        # Clean up the JSON file\n",
    "        os.remove('results.json')\n",
    "    else:\n",
    "        # Calculate the statistics for the entire dataset\n",
    "        p = data[['methylated', 'unmethylated']].values.flatten().astype(np.float64)\n",
    "        q = data[['unmethylated', 'methylated']].values.flatten().astype(np.float64)\n",
    "\n",
    "        # Apply smoothing\n",
    "        p += smoothing\n",
    "        q += smoothing\n",
    "\n",
    "        # Normalize p and q\n",
    "        p /= p.sum()\n",
    "        q /= q.sum()\n",
    "\n",
    "        logging.info(f\"HERE IS THE P: {p}\")\n",
    "        logging.info(f\"HERE IS THE Q: {q}\")\n",
    "        \n",
    "\n",
    "        kl = kl_divergence(p, q)\n",
    "        js = jensenshannon(p, q)\n",
    "        gjs = geometric_jsd(p, q)\n",
    "        ks = ks_test(p, q)\n",
    "\n",
    "        dataset_result = {\n",
    "            'entropy': entropy(p),\n",
    "            'relative_entropy': kl,  \n",
    "            'dkl_divergence': kl,\n",
    "            'jsd': js,\n",
    "            'geometric_jsd': gjs,\n",
    "            'kolmogorov_smirnov': ks\n",
    "        }\n",
    "\n",
    "        logging.info(f\"Reult: {kl, js, gjs, ks}\")\n",
    "        \n",
    "        \n",
    "        return dataset_result\n",
    "\n",
    "\n",
    "def read_bismark_file(filename):\n",
    "    \"\"\"Reads Bismark methylation data from a file.\"\"\"\n",
    "    column_names = [\"chr\", \"start\", \"end\", \"coverage\", \"methylated\", \"unmethylated\"]\n",
    "    df = pd.read_csv(filename, sep='\\t', header=None, names=column_names, compression='gzip')\n",
    "\n",
    "    # Convert columns to correct data types\n",
    "    df['start'] = pd.to_numeric(df['start'])\n",
    "    df['end'] = pd.to_numeric(df['end'])\n",
    "    df['coverage'] = pd.to_numeric(df['coverage'])\n",
    "    df['methylated'] = pd.to_numeric(df['methylated'])\n",
    "    df['unmethylated'] = pd.to_numeric(df['unmethylated'])\n",
    "\n",
    "    return df\n",
    "\n",
    "def clean_data(results_df):\n",
    "    \"\"\"Cleans the results DataFrame by removing file extensions from sample names.\"\"\"\n",
    "    results_df['Sample'] = results_df['Sample'].astype(str).str.replace('.bedgraph.gz', '', regex=False)\n",
    "    return results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a91a5aea-7165-4843-bef7-aecb7f5dc10a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-24T15:56:56.519597Z",
     "iopub.status.busy": "2024-06-24T15:56:56.518813Z",
     "iopub.status.idle": "2024-06-24T15:56:56.525888Z",
     "shell.execute_reply": "2024-06-24T15:56:56.525247Z",
     "shell.execute_reply.started": "2024-06-24T15:56:56.519578Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# Data processing pipeline\n",
    "data_directory = \"/home/eharpu/methylation_analysis/samples_testing\"\n",
    "results = []\n",
    "\n",
    "# Get all sample subdirectories\n",
    "#sample_directories = [d for d in os.listdir(data_directory) if os.path.isdir(os.path.join(data_directory, d))]\n",
    "\n",
    "# Iterate through files\n",
    "for sample_dir in [\"CHH_SRX1664497\"]:\n",
    "    sample_path = os.path.join(data_directory, sample_dir)\n",
    "\n",
    "    logging.info(f\"Processing sample directory: {sample_path}\")\n",
    "\n",
    "    # Iterate through files in each sample directory\n",
    "    for file_path in glob.glob(os.path.join(sample_path, \"*.cov.gz\")):\n",
    "        try:\n",
    "            filename = os.path.basename(file_path)\n",
    "            parts = filename.split(\"_\")\n",
    "\n",
    "            context_type = parts[0]\n",
    "            sample_name = parts[1] + \"_\" + parts[2].split(\".\")[0] \n",
    "            chromosome = parts[-1].split(\".\")[0]\n",
    "\n",
    "            logging.info(f\"Processing file: {filename}\")\n",
    "\n",
    "            df = read_bismark_file(file_path)\n",
    "            divergences = divergence_calculations(df, False)\n",
    "\n",
    "            logging.info(f\"divergence calculated: {filename}\")\n",
    "            \n",
    "            results.append(\n",
    "                {\n",
    "                    \"Sample\": sample_name,\n",
    "                    \"Chromosome\": chromosome,\n",
    "                    \"Context\": context_type,\n",
    "                    \"JS Divergence\": divergences[0],  \n",
    "                    \"KL Divergence\": divergences[1],\n",
    "                    \"GJS Divergence\": divergences[2],\n",
    "                    \"SGJS Divergence\": divergences[3],\n",
    "                    \"KS Statistic\": divergences[4],\n",
    "                    \"KS P-value\": divergences[5]\n",
    "                }\n",
    "            )\n",
    "    \n",
    "        except Exception as e:\n",
    "            logging.error(f\"Error processing file {filename}: {e}\")\n",
    "\n",
    "# Create a DataFrame from results\n",
    "results_df = pd.DataFrame(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5405027-535c-40a1-8d19-96c7e6646eed",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df.to_csv(\"test.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e05cf3ec-d6b4-4858-8099-447473094e2a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-06-24T15:55:34.370137Z",
     "iopub.status.busy": "2024-06-24T15:55:34.369445Z",
     "iopub.status.idle": "2024-06-24T15:55:34.374284Z",
     "shell.execute_reply": "2024-06-24T15:55:34.373822Z",
     "shell.execute_reply.started": "2024-06-24T15:55:34.370114Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52cedf6e-745d-4ec3-8169-b98408e862e8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "environment_shannon",
   "language": "python",
   "name": "environment_shannon"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
